contentgrid-spring-audit-logging/src/main/java/com/contentgrid/spring/audit/handler/messaging/AuditEventToCloudEventMessageConverter.java

```diff
@@ -0,0 +1,94 @@
+import com.contentgrid.spring.audit.event.BasicAuditEvent;
+import com.contentgrid.spring.audit.event.EntityContentAuditEvent;
+import com.contentgrid.spring.audit.event.EntityItemAuditEvent;
+import com.contentgrid.spring.audit.event.EntityItemAuditEvent.Operation;
+import com.contentgrid.spring.audit.event.EntitySearchAuditEvent;
+import io.cloudevents.core.builder.CloudEventBuilder;
+import io.cloudevents.core.data.PojoCloudEventData;
+import java.net.URI;
+import java.util.Locale;
+import java.util.Set;
+import java.util.UUID;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.MessageConverter;
+
+@RequiredArgsConstructor
+public class AuditEventToCloudEventMessageConverter implements MessageConverter {
+
+    @NonNull
+    private final MessageConverter delegate;
+
+    @NonNull
+    private final PojoCloudEventData.ToBytes<AbstractAuditEvent> mapper;
+
+    @NonNull
+    private final URI source;
+
+    private static final Set<AuditEventToCloudEventMapper> TYPE_MAPPING = Set.of(
+            AuditEventToCloudEventMapper.build(BasicAuditEvent.class)
+                    .cloudEventType("cloud.contentgrid.audit.basic")
+                    .cloudEventSubject(BasicAuditEvent::getRequestUri)
+                    .build(),
+            AuditEventToCloudEventMapper.build(EntityItemAuditEvent.class)
+                    .cloudEventType("cloud.contentgrid.audit.entity",
+                            e -> e.getOperation().name().toLowerCase(Locale.ROOT))
+                    .cloudEventSubject(
+                            e -> e.getOperation() == Operation.CREATE ? e.getResponseLocation() : e.getRequestUri())
+                    .build(),
+            AuditEventToCloudEventMapper.build(AbstractEntityRelationAuditEvent.class)
+                    .cloudEventType("cloud.contentgrid.audit.entity.relation",
+                            e -> e.getOperation().name().toLowerCase(Locale.ROOT))
+                    .cloudEventSubject(AbstractAuditEvent::getRequestUri)
+                    .build(),
+            AuditEventToCloudEventMapper.build(EntityContentAuditEvent.class)
+                    .cloudEventType("cloud.contentgrid.audit.entity.content",
+                            e -> e.getOperation().name().toLowerCase(Locale.ROOT))
+                    .cloudEventSubject(AbstractAuditEvent::getRequestUri)
+                    .build(),
+            AuditEventToCloudEventMapper.build(EntitySearchAuditEvent.class)
+                    .cloudEventType("cloud.contentgrid.audit.entity.list")
+                    .cloudEventSubject(AbstractAuditEvent::getRequestUri)
+                    .build()
+    );
+
+    @Override
+    public Object fromMessage(Message<?> message, Class<?> targetClass) {
+        // TODO: implement converting back from a message to audit event?
```

Comment:
rather throw an exception, isntead of silently returning `null` ?

Reply:
When returning null, another message converter gets a chance to convert it. If no message converters return an object, an exception is thrown from somewhere.

Reply:
Ok - what should happen with this `// TODO` ? When would we need the conversion in the other direction ?